cpp/src/explainer/kernel_shap.cu

```diff
@@ -0,0 +1,207 @@
+namespace Explainer {
+
+/*
+* Kernel distrubutes exact part of the kernel shap dataset
+* Each block scatters the data of a row of `observations` into the (number of rows of
+* background) in `dataset`, based on the row of `X`.
+* So, given:
+* background = [[0, 1, 2],
+                [3, 4, 5]]
+* observation = [100, 101, 102]
+* X = [[1, 0, 1],
+*      [0, 1, 1]]
+*
+* dataset (output):
+* [[100, 1, 102],
+*  [100, 4, 102]
+*  [0, 101, 102],
+*  [3, 101, 102]]
+*
+*
+*/
+
+template <typename DataT, typename IdxT>
+__global__ void exact_rows_kernel(float* X, IdxT nrows_X, IdxT ncols,
+                                  DataT* background, IdxT nrows_background,
+                                  DataT* dataset, DataT* observation) {
+  // Each block processes one row of X. Columns are iterated over by blockDim.x at a time to ensure data coelescing
+  int col = threadIdx.x;
+  int row = blockIdx.x * ncols;
+
+  while (col < ncols) {
+    // Load the X idx for the current column
+    int curr_X = (int)X[row + col];
+
+    // Iterate over nrows_background
+    for (int row_idx = blockIdx.x * nrows_background;
+         row_idx < blockIdx.x * nrows_background + nrows_background;
+         row_idx += 1) {
+      if (curr_X == 0) {
+        dataset[row_idx * ncols + col] =
+          background[(row_idx % nrows_background) * ncols + col];
+      } else {
+        dataset[row_idx * ncols + col] = observation[col];
+      }
+    }
+    // Increment the column
+    col += blockDim.x;
+  }
+}
+
+/*
+* Kernel distrubutes sampled part of the kernel shap dataset
+* The first thread of each block calculates the sampling of `k` entries of `observation`
+* to scatter into `dataset`. Afterwards each block scatters the data of a row of `X` into the (number of rows of
+* background) in `dataset`.
+* So, given:
+* background = [[0, 1, 2, 3],
+                [5, 6, 7, 8]]
+* observation = [100, 101, 102, 103]
+* nsamples = [3, 2]
+*
+* X (output)
+*      [[1, 0, 1, 1],
+*       [0, 1, 1, 0]]
+*
+* dataset (output):
+* [[100, 1, 102, 103],
+*  [100, 6, 102, 103]
+*  [0, 101, 102, 3],
+*  [5, 101, 102, 8]]
+*
+*
+*/
+template <typename DataT, typename IdxT>
+__global__ void sampled_rows_kernel(IdxT* nsamples, float* X, IdxT nrows_X,
+                                    IdxT ncols, DataT* background,
+                                    IdxT nrows_background, DataT* dataset,
+                                    DataT* observation, uint64_t seed) {
+  int tid = threadIdx.x + blockIdx.x * blockDim.x;
+  // see what k this block will generate
+  int k_blk = nsamples[blockIdx.x];
+
+  // First k threads of block generate samples
+  if (threadIdx.x < k_blk) {
+    curandStatePhilox4_32_10_t state;
+    curand_init((unsigned long long)seed, (unsigned long long)tid, 0, &state);
+    int rand_idx = (int)(curand_uniform(&state) * ncols);
+
+    // Since X is initialized to 0, we quickly check for collisions (if k_blk << ncols the likelyhood of collisions is low)
```

Comment:
yeah, since `rand_idx * ncols` gets converted to `int`, collisions happen sometimes if `ncols` is not big (even if its big, just way more unlikely). If we didn't check, then we would generate less `1`s than we wanted, and for small `ncols` that could increase the variance in the final results